# Importing required libraries

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

# Loading Dataset

In [2]:
df=pd.read_excel(r"D:\Hackothon\uber_rides_data.xlsx")

In [3]:
df

ride_id  fare_amount          pickup_datetime  pickup_longitude   
0       24238194          7.5  2015-05-07 19:52:06 UTC        -73.999817  \
1       27835199          7.7  2009-07-17 20:04:56 UTC        -73.994355   
2       44984355         12.9  2009-08-24 21:45:00 UTC        -74.005043   
3       25894730          5.3  2009-06-26 08:22:21 UTC        -73.976124   
4       17610152         16.0  2014-08-28 17:47:00 UTC        -73.925023   
...          ...          ...                      ...               ...   
199995  42598914          3.0  2012-10-28 10:49:00 UTC        -73.987042   
199996  16382965          7.5  2014-03-14 01:09:00 UTC        -73.984722   
199997  27804658         30.9  2009-06-29 00:42:00 UTC        -73.986017   
199998  20259894         14.5  2015-05-20 14:56:25 UTC        -73.997124   
199999  11951496         14.1  2010-05-15 04:08:00 UTC        -73.984395   

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  
0             40.738354         -73.999512         40.723217                1  
1             40.728225         -73.994710         40.750325                1  
2             40.740770         -73.962565         40.772647                1  
3             40.790844         -73.965316         40.803349                3  
4             40.744085         -73.973082         40.761247                5  
...                 ...                ...               ...              ...  
199995        40.739367         -73.986525         40.740297                1  
199996        40.736837         -74.006672         40.739620                1  
199997        40.756487         -73.858957         40.692588                2  
199998        40.725452         -73.983215         40.695415                1  
199999        40.720077         -73.985508         40.768793                1  

[200000 rows x 8 columns]

# shape of dataset

In [4]:
df.shape

(200000, 8)

# Number of integer columns given in dataset

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Data columns (total 8 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   ride_id            200000 non-null  int64  
 1   fare_amount        200000 non-null  float64
 2   pickup_datetime    200000 non-null  object 
 3   pickup_longitude   200000 non-null  float64
 4   pickup_latitude    200000 non-null  float64
 5   dropoff_longitude  199999 non-null  float64
 6   dropoff_latitude   199999 non-null  float64
 7   passenger_count    200000 non-null  int64  
dtypes: float64(5), int64(2), object(1)
memory usage: 12.2+ MB


### There are two integer columns in datset
### first column is ride_id
### second column is passenger_count 

# Number of missing values in 'dropoff_longitude' column

In [6]:
df['dropoff_longitude'].isnull().sum()

1

In [7]:
null_data = df[df.isnull().any(axis=1)]

In [8]:
null_data

ride_id  fare_amount          pickup_datetime  pickup_longitude   
87946  32736015         24.1  2013-07-02 03:51:57 UTC        -73.950581  \

       pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  
87946        40.779692                NaN               NaN                0

### By the above statement we can obsserve there are one missing value in 'dropoff_longitude' one in 'dropoff_latitude'.

# The Datatype of 'pickup_datetime' feature

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Data columns (total 8 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   ride_id            200000 non-null  int64  
 1   fare_amount        200000 non-null  float64
 2   pickup_datetime    200000 non-null  object 
 3   pickup_longitude   200000 non-null  float64
 4   pickup_latitude    200000 non-null  float64
 5   dropoff_longitude  199999 non-null  float64
 6   dropoff_latitude   199999 non-null  float64
 7   passenger_count    200000 non-null  int64  
dtypes: float64(5), int64(2), object(1)
memory usage: 12.2+ MB


In [10]:
df.dtypes['pickup_datetime']

dtype('O')

### from the above statement we can say that the datatype of pickup_datetime is "object"

# correct syntax to convert 'pickup_datetime' to datetime datatype is

In [14]:
df['pickup_datetime']=pd.to_datetime(df['pickup_datetime'])

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Data columns (total 8 columns):
 #   Column             Non-Null Count   Dtype              
---  ------             --------------   -----              
 0   ride_id            200000 non-null  int64              
 1   fare_amount        200000 non-null  float64            
 2   pickup_datetime    200000 non-null  datetime64[ns, UTC]
 3   pickup_longitude   200000 non-null  float64            
 4   pickup_latitude    200000 non-null  float64            
 5   dropoff_longitude  199999 non-null  float64            
 6   dropoff_latitude   199999 non-null  float64            
 7   passenger_count    200000 non-null  int64              
dtypes: datetime64[ns, UTC](1), float64(5), int64(2)
memory usage: 12.2 MB


### only pd.to_datetime method is working for this data and it is the correct syntax to convert to datetime datatype

# To remove null values from dataframe

In [16]:
df.isnull().sum()

ride_id              0
fare_amount          0
pickup_datetime      0
pickup_longitude     0
pickup_latitude      0
dropoff_longitude    1
dropoff_latitude     1
passenger_count      0
dtype: int64

In [19]:
df=df.dropna()

In [20]:
df.isnull().sum()

ride_id              0
fare_amount          0
pickup_datetime      0
pickup_longitude     0
pickup_latitude      0
dropoff_longitude    0
dropoff_latitude     0
passenger_count      0
dtype: int64

# Average fare amount after removing null values from dataframe

In [21]:
Average_fareamount=df['fare_amount'].mean()

In [22]:
print("Average_fare_amount",Average_fareamount)

Average_fare_amount 11.359891549457748


# Median havesine distance between pickup and dropoff location 

In [23]:
import math

In [26]:
# creating function for haversine distance 
def haversine_distance(latitude1, longitude1, latitude2, longitude2):
    #readius of earth in kms
    radius = 6371.0

    dlat = np.radians(latitude2 - latitude1)
    

    dlon = np.radians(longitude2 - longitude1)
    
    # Haversine formula
    a = np.sin(dlat / 2)**2 + np.cos(np.radians(latitude1)) * np.cos(np.radians(latitude2)) * np.sin(dlon / 2)**2
    c = 2 * np.arctan2(np.sqrt(a), np.sqrt(1 - a))

    distance = radius * c
    
    return distance

In [27]:
df['haversine_distance'] = df.apply(lambda row: haversine_distance(row['pickup_latitude'], row['pickup_longitude'], row['dropoff_latitude'], row['dropoff_longitude']), axis=1)

C:\Users\user\AppData\Local\Temp\ipykernel_7808\433663663.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['haversine_distance'] = df.apply(lambda row: haversine_distance(row['pickup_latitude'], row['pickup_longitude'], row['dropoff_latitude'], row['dropoff_longitude']), axis=1)


In [28]:
df

ride_id  fare_amount           pickup_datetime  pickup_longitude   
0       24238194          7.5 2015-05-07 19:52:06+00:00        -73.999817  \
1       27835199          7.7 2009-07-17 20:04:56+00:00        -73.994355   
2       44984355         12.9 2009-08-24 21:45:00+00:00        -74.005043   
3       25894730          5.3 2009-06-26 08:22:21+00:00        -73.976124   
4       17610152         16.0 2014-08-28 17:47:00+00:00        -73.925023   
...          ...          ...                       ...               ...   
199995  42598914          3.0 2012-10-28 10:49:00+00:00        -73.987042   
199996  16382965          7.5 2014-03-14 01:09:00+00:00        -73.984722   
199997  27804658         30.9 2009-06-29 00:42:00+00:00        -73.986017   
199998  20259894         14.5 2015-05-20 14:56:25+00:00        -73.997124   
199999  11951496         14.1 2010-05-15 04:08:00+00:00        -73.984395   

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count   
0             40.738354         -73.999512         40.723217                1  \
1             40.728225         -73.994710         40.750325                1   
2             40.740770         -73.962565         40.772647                1   
3             40.790844         -73.965316         40.803349                3   
4             40.744085         -73.973082         40.761247                5   
...                 ...                ...               ...              ...   
199995        40.739367         -73.986525         40.740297                1   
199996        40.736837         -74.006672         40.739620                1   
199997        40.756487         -73.858957         40.692588                2   
199998        40.725452         -73.983215         40.695415                1   
199999        40.720077         -73.985508         40.768793                1   

        haversine_distance  
0                 1.683323  
1                 2.457590  
2                 5.036377  
3                 1.661683  
4                 4.475450  
...                    ...  
199995            0.112210  
199996            1.875050  
199997           12.850319  
199998            3.539715  
199999            5.417783  

[199999 rows x 9 columns]

In [29]:
median_haversine_distance = df['haversine_distance'].median()

print("Median Haversine Distance:", median_haversine_distance)

Median Haversine Distance: 2.120992396183407


## Median Haversine Distance: 2.120992396183407

# Maximum haversine distance between pickup and dropoff location according to the given dataset

In [30]:
# Find the maximum Haversine distance
max_haversine_distance = df['haversine_distance'].max()

print("Maximum Haversine Distance:", max_haversine_distance)

Maximum Haversine Distance: 16409.239135313186


### Maximum haversine distance for pickup and dropoff location for dataset is 16409.239135313186

# Number of rides have 0.0 haversine distance between pickup and dropoff location according to the given dataset

In [32]:
df['haversine_distance'].value_counts().get(0.0,0)

5632

In [40]:
distance_counts = df['haversine_distance'].value_counts()

rides_with_zero_distance = distance_counts.get(0.0, 0)

print("Number of rides with 0.0 Haversine distance:", rides_with_zero_distance)

Number of rides with 0.0 Haversine distance: 5632


# mean 'fare_amount' for rides with 0 haversine distance

In [42]:
zero_distance = df[df['haversine_distance'] == 0.0]

mean_fare_for_zero_distance = zero_distance['fare_amount'].mean()

print("Mean 'fare_amount' for rides with 0 Haversine distance:", mean_fare_for_zero_distance)

Mean 'fare_amount' for rides with 0 Haversine distance: 11.585317826704546


In [43]:
# Maximum 'fare_amount' for a ride

In [45]:
maximum_fare_for_ride=df['fare_amount'].max()

print('maximum_fare_for_ride',maximum_fare_for_ride)

maximum_fare_for_ride 499.0


# The haversine distance between pickup and dropoff location for the costliest ride

In [46]:
max_fare=df['fare_amount'].max()
max_fare

499.0

In [49]:
costly_fare=df[df['fare_amount']==max_fare]
costly_fare

ride_id  fare_amount           pickup_datetime  pickup_longitude   
170081  51151143        499.0 2011-04-10 04:10:00+00:00        -73.968377  \

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count   
170081        40.764602         -73.968368           40.7646                1  \

        haversine_distance  
170081             0.00079

In [55]:
haversine_distance_costly_fare=costly_fare['haversine_distance'].values[0]

In [56]:
print('haversine_distance_costly_fare',haversine_distance_costly_fare)

haversine_distance_costly_fare 0.0007899213197627205


# Number of rides recorded in the year 2014

In [62]:
df['pickup_year']=df['pickup_datetime'].dt.year
df

C:\Users\user\AppData\Local\Temp\ipykernel_7808\1407686941.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['pickup_year']=df['pickup_datetime'].dt.year


ride_id  fare_amount           pickup_datetime  pickup_longitude   
0       24238194          7.5 2015-05-07 19:52:06+00:00        -73.999817  \
1       27835199          7.7 2009-07-17 20:04:56+00:00        -73.994355   
2       44984355         12.9 2009-08-24 21:45:00+00:00        -74.005043   
3       25894730          5.3 2009-06-26 08:22:21+00:00        -73.976124   
4       17610152         16.0 2014-08-28 17:47:00+00:00        -73.925023   
...          ...          ...                       ...               ...   
199995  42598914          3.0 2012-10-28 10:49:00+00:00        -73.987042   
199996  16382965          7.5 2014-03-14 01:09:00+00:00        -73.984722   
199997  27804658         30.9 2009-06-29 00:42:00+00:00        -73.986017   
199998  20259894         14.5 2015-05-20 14:56:25+00:00        -73.997124   
199999  11951496         14.1 2010-05-15 04:08:00+00:00        -73.984395   

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count   
0             40.738354         -73.999512         40.723217                1  \
1             40.728225         -73.994710         40.750325                1   
2             40.740770         -73.962565         40.772647                1   
3             40.790844         -73.965316         40.803349                3   
4             40.744085         -73.973082         40.761247                5   
...                 ...                ...               ...              ...   
199995        40.739367         -73.986525         40.740297                1   
199996        40.736837         -74.006672         40.739620                1   
199997        40.756487         -73.858957         40.692588                2   
199998        40.725452         -73.983215         40.695415                1   
199999        40.720077         -73.985508         40.768793                1   

        haversine_distance  pickup_year  
0                 1.683323         2015  
1                 2.457590         2009  
2                 5.036377         2009  
3                 1.661683         2009  
4                 4.475450         2014  
...                    ...          ...  
199995            0.112210         2012  
199996            1.875050         2014  
199997           12.850319         2009  
199998            3.539715         2015  
199999            5.417783         2010  

[199999 rows x 10 columns]

In [63]:
df[df['pickup_year']==2014]

ride_id  fare_amount           pickup_datetime  pickup_longitude   
4       17610152        16.00 2014-08-28 17:47:00+00:00        -73.925023  \
6       48725865        24.50 2014-10-12 07:04:00+00:00        -73.961447   
20      55085966        10.50 2014-02-18 14:26:00+00:00        -73.980022   
26      38755863         5.00 2014-01-21 06:55:00+00:00        -73.957802   
34      19277743        39.50 2014-06-04 06:49:00+00:00        -73.788080   
...          ...          ...                       ...               ...   
199979   9699676         6.00 2014-06-08 21:38:43+00:00        -73.975209   
199981  21553740         9.00 2014-06-02 21:51:50+00:00        -73.980445   
199982  13096190        57.33 2014-08-06 11:06:06+00:00        -73.969204   
199994   3189201        12.00 2014-01-31 14:42:00+00:00        -73.983070   
199996  16382965         7.50 2014-03-14 01:09:00+00:00        -73.984722   

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count   
4             40.744085         -73.973082         40.761247                5  \
6             40.693965         -73.871195         40.774297                5   
20            40.745990         -74.003432         40.759667                1   
26            40.776372         -73.957422         40.782870                1   
34            40.642187         -73.865042         40.725997                4   
...                 ...                ...               ...              ...   
199979        40.752767         -73.963186         40.757905                2   
199981        40.761823         -73.954004         40.784407                1   
199982        40.754771         -73.790351         40.643802                1   
199994        40.760770         -73.972972         40.754177                1   
199996        40.736837         -74.006672         40.739620                1   

        haversine_distance  pickup_year  
4                 4.475450         2014  
6                11.731015         2014  
20                2.490244         2014  
26                0.723253         2014  
34               11.356141         2014  
...                    ...          ...  
199979            1.162745         2014  
199981            3.356153         2014  
199982           19.483047         2014  
199994            1.122878         2014  
199996            1.875050         2014  

[29968 rows x 10 columns]

In [66]:
count_of_rides_in_2014=df[df['pickup_year'] == 2014].shape[0]
print('count_of_rides_in_2014=',count_of_rides_in_2014)

count_of_rides_in_2014= 29968


# Number of rides were recorded in the first quarter of 2014

In [68]:
df['pickup_month'] = df['pickup_datetime'].dt.month
df

C:\Users\user\AppData\Local\Temp\ipykernel_7808\4155855579.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['pickup_month'] = df['pickup_datetime'].dt.month


ride_id  fare_amount           pickup_datetime  pickup_longitude   
0       24238194          7.5 2015-05-07 19:52:06+00:00        -73.999817  \
1       27835199          7.7 2009-07-17 20:04:56+00:00        -73.994355   
2       44984355         12.9 2009-08-24 21:45:00+00:00        -74.005043   
3       25894730          5.3 2009-06-26 08:22:21+00:00        -73.976124   
4       17610152         16.0 2014-08-28 17:47:00+00:00        -73.925023   
...          ...          ...                       ...               ...   
199995  42598914          3.0 2012-10-28 10:49:00+00:00        -73.987042   
199996  16382965          7.5 2014-03-14 01:09:00+00:00        -73.984722   
199997  27804658         30.9 2009-06-29 00:42:00+00:00        -73.986017   
199998  20259894         14.5 2015-05-20 14:56:25+00:00        -73.997124   
199999  11951496         14.1 2010-05-15 04:08:00+00:00        -73.984395   

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count   
0             40.738354         -73.999512         40.723217                1  \
1             40.728225         -73.994710         40.750325                1   
2             40.740770         -73.962565         40.772647                1   
3             40.790844         -73.965316         40.803349                3   
4             40.744085         -73.973082         40.761247                5   
...                 ...                ...               ...              ...   
199995        40.739367         -73.986525         40.740297                1   
199996        40.736837         -74.006672         40.739620                1   
199997        40.756487         -73.858957         40.692588                2   
199998        40.725452         -73.983215         40.695415                1   
199999        40.720077         -73.985508         40.768793                1   

        haversine_distance  pickup_year  pickup_month  
0                 1.683323         2015             5  
1                 2.457590         2009             7  
2                 5.036377         2009             8  
3                 1.661683         2009             6  
4                 4.475450         2014             8  
...                    ...          ...           ...  
199995            0.112210         2012            10  
199996            1.875050         2014             3  
199997           12.850319         2009             6  
199998            3.539715         2015             5  
199999            5.417783         2010             5  

[199999 rows x 11 columns]

In [69]:
count_of_rides_in_2014_in_first_quarter = df[(df['pickup_year'] == 2014) & (df['pickup_month'] >= 1) & (df['pickup_month'] <= 3)].shape[0]
print('count_of_rides_in_2014_in_first_quarter',count_of_rides_in_2014_in_first_quarter)

count_of_rides_in_2014_in_first_quarter 7687


# Day of the week in September 2010, maximum rides were recorded ?

In [72]:
df['day_of_week'] = df['pickup_datetime'].dt.day_name()
df

C:\Users\user\AppData\Local\Temp\ipykernel_7808\799383257.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['day_of_week'] = df['pickup_datetime'].dt.day_name()


ride_id  fare_amount           pickup_datetime  pickup_longitude   
0       24238194          7.5 2015-05-07 19:52:06+00:00        -73.999817  \
1       27835199          7.7 2009-07-17 20:04:56+00:00        -73.994355   
2       44984355         12.9 2009-08-24 21:45:00+00:00        -74.005043   
3       25894730          5.3 2009-06-26 08:22:21+00:00        -73.976124   
4       17610152         16.0 2014-08-28 17:47:00+00:00        -73.925023   
...          ...          ...                       ...               ...   
199995  42598914          3.0 2012-10-28 10:49:00+00:00        -73.987042   
199996  16382965          7.5 2014-03-14 01:09:00+00:00        -73.984722   
199997  27804658         30.9 2009-06-29 00:42:00+00:00        -73.986017   
199998  20259894         14.5 2015-05-20 14:56:25+00:00        -73.997124   
199999  11951496         14.1 2010-05-15 04:08:00+00:00        -73.984395   

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count   
0             40.738354         -73.999512         40.723217                1  \
1             40.728225         -73.994710         40.750325                1   
2             40.740770         -73.962565         40.772647                1   
3             40.790844         -73.965316         40.803349                3   
4             40.744085         -73.973082         40.761247                5   
...                 ...                ...               ...              ...   
199995        40.739367         -73.986525         40.740297                1   
199996        40.736837         -74.006672         40.739620                1   
199997        40.756487         -73.858957         40.692588                2   
199998        40.725452         -73.983215         40.695415                1   
199999        40.720077         -73.985508         40.768793                1   

        haversine_distance  pickup_year  pickup_month day_of_week  
0                 1.683323         2015             5    Thursday  
1                 2.457590         2009             7      Friday  
2                 5.036377         2009             8      Monday  
3                 1.661683         2009             6      Friday  
4                 4.475450         2014             8    Thursday  
...                    ...          ...           ...         ...  
199995            0.112210         2012            10      Sunday  
199996            1.875050         2014             3      Friday  
199997           12.850319         2009             6      Monday  
199998            3.539715         2015             5   Wednesday  
199999            5.417783         2010             5    Saturday  

[199999 rows x 12 columns]

In [74]:
september_2010_rides = df[(df['pickup_year'] == 2010) & (df['pickup_datetime'].dt.month == 9)]
september_2010_rides

ride_id  fare_amount           pickup_datetime  pickup_longitude   
23      25121708          7.7 2010-09-04 16:12:00+00:00        -73.994300  \
85      11536406          7.7 2010-09-08 10:16:01+00:00        -73.987383   
87      11267091          4.5 2010-09-22 12:15:00+00:00        -73.981920   
126      8791172          5.3 2010-09-21 12:38:00+00:00        -73.956158   
129     30203807         10.5 2010-09-10 16:28:00+00:00        -73.995627   
...          ...          ...                       ...               ...   
199584  52152685          4.5 2010-09-02 02:07:00+00:00        -73.995380   
199665  44243859          8.5 2010-09-04 21:05:00+00:00        -74.001550   
199851  37613854         11.7 2010-09-13 14:46:00+00:00        -74.009923   
199862   5095334          6.9 2010-09-30 11:29:20+00:00        -73.976584   
199901  19337978          8.9 2010-09-22 08:19:31+00:00        -73.930492   

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count   
23            40.739512         -73.988070         40.724482                2  \
85            40.721255         -73.954733         40.730683                1   
87            40.761172         -74.016152         40.704765                1   
126           40.781593         -73.966138         40.773645                1   
129           40.749295         -73.965087         40.771610                1   
...                 ...                ...               ...              ...   
199584        40.721638         -73.988308         40.731667                1   
199665        40.741060         -73.995810         40.739425                3   
199851        40.714860         -74.007020         40.705720                1   
199862        40.757000         -73.991153         40.765653                1   
199901        40.761509         -73.963917         40.759702                1   

        haversine_distance  pickup_year  pickup_month day_of_week  
23                1.751763         2010             9    Saturday  
85                2.944304         2010             9   Wednesday  
87                6.903596         2010             9   Wednesday  
126               1.219522         2010             9     Tuesday  
129               3.573956         2010             9      Friday  
...                    ...          ...           ...         ...  
199584            1.264418         2010             9    Thursday  
199665            0.516638         2010             9    Saturday  
199851            1.045362         2010             9      Monday  
199862            1.559298         2010             9    Thursday  
199901            2.822347         2010             9   Wednesday  

[2482 rows x 12 columns]

In [75]:
rides_by_day_of_week = september_2010_rides['day_of_week'].value_counts()
rides_by_day_of_week

day_of_week
Thursday     457
Wednesday    391
Saturday     362
Friday       354
Sunday       331
Tuesday      322
Monday       265
Name: count, dtype: int64

In [80]:
max_rides_day = rides_by_day_of_week.idxmax()
print("max_rides_day:",max_rides_day)

max_rides_day: Thursday


# Apply a Machine Learning Algorithm to predict the fare amount given following input features: passenger_count, distance and ride_week_day
# Perform a 70-30 split of data
# Which algorithm gives the least adjusted R square value

In [81]:
# importing required methods,clases algorithms and metrics

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import r2_score

In [84]:
df

ride_id  fare_amount           pickup_datetime  pickup_longitude   
0       24238194          7.5 2015-05-07 19:52:06+00:00        -73.999817  \
1       27835199          7.7 2009-07-17 20:04:56+00:00        -73.994355   
2       44984355         12.9 2009-08-24 21:45:00+00:00        -74.005043   
3       25894730          5.3 2009-06-26 08:22:21+00:00        -73.976124   
4       17610152         16.0 2014-08-28 17:47:00+00:00        -73.925023   
...          ...          ...                       ...               ...   
199995  42598914          3.0 2012-10-28 10:49:00+00:00        -73.987042   
199996  16382965          7.5 2014-03-14 01:09:00+00:00        -73.984722   
199997  27804658         30.9 2009-06-29 00:42:00+00:00        -73.986017   
199998  20259894         14.5 2015-05-20 14:56:25+00:00        -73.997124   
199999  11951496         14.1 2010-05-15 04:08:00+00:00        -73.984395   

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count   
0             40.738354         -73.999512         40.723217                1  \
1             40.728225         -73.994710         40.750325                1   
2             40.740770         -73.962565         40.772647                1   
3             40.790844         -73.965316         40.803349                3   
4             40.744085         -73.973082         40.761247                5   
...                 ...                ...               ...              ...   
199995        40.739367         -73.986525         40.740297                1   
199996        40.736837         -74.006672         40.739620                1   
199997        40.756487         -73.858957         40.692588                2   
199998        40.725452         -73.983215         40.695415                1   
199999        40.720077         -73.985508         40.768793                1   

        haversine_distance  pickup_year  pickup_month day_of_week  
0                 1.683323         2015             5    Thursday  
1                 2.457590         2009             7      Friday  
2                 5.036377         2009             8      Monday  
3                 1.661683         2009             6      Friday  
4                 4.475450         2014             8    Thursday  
...                    ...          ...           ...         ...  
199995            0.112210         2012            10      Sunday  
199996            1.875050         2014             3      Friday  
199997           12.850319         2009             6      Monday  
199998            3.539715         2015             5   Wednesday  
199999            5.417783         2010             5    Saturday  

[199999 rows x 12 columns]

In [89]:
X = df[['passenger_count', 'haversine_distance', 'day_of_week']]
y = df['fare_amount']

In [90]:
X

passenger_count  haversine_distance day_of_week
0                     1            1.683323    Thursday
1                     1            2.457590      Friday
2                     1            5.036377      Monday
3                     3            1.661683      Friday
4                     5            4.475450    Thursday
...                 ...                 ...         ...
199995                1            0.112210      Sunday
199996                1            1.875050      Friday
199997                2           12.850319      Monday
199998                1            3.539715   Wednesday
199999                1            5.417783    Saturday

[199999 rows x 3 columns]

In [91]:
y

0          7.5
1          7.7
2         12.9
3          5.3
4         16.0
          ... 
199995     3.0
199996     7.5
199997    30.9
199998    14.5
199999    14.1
Name: fare_amount, Length: 199999, dtype: float64

In [92]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [94]:
X_train.shape

(139999, 3)

In [95]:
X_test.shape

(60000, 3)

In [96]:
y_train.shape

(139999,)

In [97]:
y_test.shape

(60000,)

In [101]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 139999 entries, 194320 to 121959
Data columns (total 3 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   passenger_count     139999 non-null  int64  
 1   haversine_distance  139999 non-null  float64
 2   day_of_week         139999 non-null  object 
dtypes: float64(1), int64(1), object(1)
memory usage: 4.3+ MB


In [100]:
X_train_cat = X_train.select_dtypes(include=['object'])

X_train_cat.head()

day_of_week
194320    Thursday
774       Saturday
75676    Wednesday
37127       Friday
96757       Sunday

In [102]:
X_train_num = X_train.select_dtypes(include=['int64', 'float64'])

X_train_num.head()

passenger_count  haversine_distance
194320                1            1.270348
774                   2            6.406598
75676                 1            3.016657
37127                 1            1.546083
96757                 5            2.492464

In [103]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_num_rescaled = pd.DataFrame(scaler.fit_transform(X_train_num), 
                                    columns = X_train_num.columns, 
                                    index = X_train_num.index)

X_train_num_rescaled.head()

passenger_count  haversine_distance
194320        -0.483094           -0.051319
774            0.221215           -0.037987
75676         -0.483094           -0.046786
37127         -0.483094           -0.050603
96757          2.334142           -0.048147

In [105]:
from sklearn.preprocessing import OneHotEncoder

encoder_ = OneHotEncoder(drop='first', sparse=False)

X_train_cat_ohe = pd.DataFrame(encoder_.fit_transform(X_train_cat), 
                               columns=encoder_.get_feature_names_out(X_train_cat.columns), 
                               index = X_train_cat.index)

X_train_cat_ohe.head()

day_of_week_Monday  day_of_week_Saturday  day_of_week_Sunday   
194320                 0.0                   0.0                 0.0  \
774                    0.0                   1.0                 0.0   
75676                  0.0                   0.0                 0.0   
37127                  0.0                   0.0                 0.0   
96757                  0.0                   0.0                 1.0   

        day_of_week_Thursday  day_of_week_Tuesday  day_of_week_Wednesday  
194320                   1.0                  0.0                    0.0  
774                      0.0                  0.0                    0.0  
75676                    0.0                  0.0                    1.0  
37127                    0.0                  0.0                    0.0  
96757                    0.0                  0.0                    0.0

In [106]:
X_test_cat = X_test.select_dtypes(include=['object'])

X_test_cat.head()

day_of_week
134253   Wednesday
124141      Sunday
62810     Saturday
65425       Sunday
30074       Monday

In [107]:
X_test_num = X_test.select_dtypes(include=['int64', 'float64'])

X_test_num.head()

passenger_count  haversine_distance
134253                1            2.098542
124141                1            0.960293
62810                 5            3.377763
65425                 1            2.716637
30074                 1            0.784348

In [108]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_test_num_rescaled = pd.DataFrame(scaler.fit_transform(X_test_num), 
                                    columns = X_test_num.columns, 
                                    index = X_test_num.index)

X_test_num_rescaled.head()

passenger_count  haversine_distance
134253        -0.522662           -0.048530
124141        -0.522662           -0.051545
62810          2.545729           -0.045142
65425         -0.522662           -0.046893
30074         -0.522662           -0.052011

In [109]:
from sklearn.preprocessing import OneHotEncoder

encoder_ = OneHotEncoder(drop='first', sparse=False)

X_test_cat_ohe = pd.DataFrame(encoder_.fit_transform(X_test_cat), 
                               columns=encoder_.get_feature_names_out(X_test_cat.columns), 
                               index = X_test_cat.index)

X_test_cat_ohe.head()

day_of_week_Monday  day_of_week_Saturday  day_of_week_Sunday   
134253                 0.0                   0.0                 0.0  \
124141                 0.0                   0.0                 1.0   
62810                  0.0                   1.0                 0.0   
65425                  0.0                   0.0                 1.0   
30074                  1.0                   0.0                 0.0   

        day_of_week_Thursday  day_of_week_Tuesday  day_of_week_Wednesday  
134253                   0.0                  0.0                    1.0  
124141                   0.0                  0.0                    0.0  
62810                    0.0                  0.0                    0.0  
65425                    0.0                  0.0                    0.0  
30074                    0.0                  0.0                    0.0

In [110]:
X_train_transformed = pd.concat([X_train_num_rescaled, X_train_cat_ohe], axis=1)

X_train_transformed.head()

passenger_count  haversine_distance  day_of_week_Monday   
194320        -0.483094           -0.051319                 0.0  \
774            0.221215           -0.037987                 0.0   
75676         -0.483094           -0.046786                 0.0   
37127         -0.483094           -0.050603                 0.0   
96757          2.334142           -0.048147                 0.0   

        day_of_week_Saturday  day_of_week_Sunday  day_of_week_Thursday   
194320                   0.0                 0.0                   1.0  \
774                      1.0                 0.0                   0.0   
75676                    0.0                 0.0                   0.0   
37127                    0.0                 0.0                   0.0   
96757                    0.0                 1.0                   0.0   

        day_of_week_Tuesday  day_of_week_Wednesday  
194320                  0.0                    0.0  
774                     0.0                    0.0  
75676                   0.0                    1.0  
37127                   0.0                    0.0  
96757                   0.0                    0.0

In [111]:
X_test_transformed = pd.concat([X_test_num_rescaled, X_test_cat_ohe], axis=1)

X_test_transformed.head()

passenger_count  haversine_distance  day_of_week_Monday   
134253        -0.522662           -0.048530                 0.0  \
124141        -0.522662           -0.051545                 0.0   
62810          2.545729           -0.045142                 0.0   
65425         -0.522662           -0.046893                 0.0   
30074         -0.522662           -0.052011                 1.0   

        day_of_week_Saturday  day_of_week_Sunday  day_of_week_Thursday   
134253                   0.0                 0.0                   0.0  \
124141                   0.0                 1.0                   0.0   
62810                    1.0                 0.0                   0.0   
65425                    0.0                 1.0                   0.0   
30074                    0.0                 0.0                   0.0   

        day_of_week_Tuesday  day_of_week_Wednesday  
134253                  0.0                    1.0  
124141                  0.0                    0.0  
62810                   0.0                    0.0  
65425                   0.0                    0.0  
30074                   0.0                    0.0

In [98]:
models = {
    'Linear Regression': LinearRegression(),
    'Decision Tree Regression': DecisionTreeRegressor(),
    'Random Forest Regression': RandomForestRegressor(),
    'K-Nearest Neighbors Regression': KNeighborsRegressor()
}

In [114]:
adjusted_r_squared_values = {}

for model_name, model in models.items():
    model.fit(X_train_transformed, y_train)
    y_pred = model.predict(X_test_transformed)
    
    # Calculate the R-squared value
    r_squared = r2_score(y_test, y_pred)
    
    # Calculate the adjusted R-squared value
    n = X_test_transformed.shape[0]  # Number of samples
    p = X_test_transformed.shape[1]  # Number of predictors (features)
    adjusted_r_squared = 1 - ((1 - r_squared) * (n - 1) / (n - p - 1))
    
    adjusted_r_squared_values[model_name] = adjusted_r_squared

In [115]:
min_adj_r_squared_model = min(adjusted_r_squared_values, key=adjusted_r_squared_values.get)
min_adj_r_squared_value = adjusted_r_squared_values[min_adj_r_squared_model]


In [116]:
print("Adjusted R-squared values for different models on the test data:")

Adjusted R-squared values for different models on the test data:


In [119]:
for model_name, adj_r_squared in adjusted_r_squared_values.items():
    print(model_name,":", adj_r_squared)

Linear Regression : 0.0006011927273995576
Decision Tree Regression : 0.464488133427732
Random Forest Regression : 0.5979487545973016
K-Nearest Neighbors Regression : 0.6080884698808027


In [120]:
print("The model with the least adjusted R-squared value on the test data is ", min_adj_r_squared_model," with a value of ", min_adj_r_squared_value,":")


The model with the least adjusted R-squared value on the test data is  Linear Regression  with a value of  0.0006011927273995576 :
